### PR TITLE
refactor(storage): refactor compaction event loop

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

https://github.com/risingwavelabs/risingwave/issues/13879

 Introduce a dedicated loop for CPU-bound / IO-bound events to avoid slow events from occupying the event loop thread for a long time, causing the basic event loop to starve.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by Little-Wallace — COMMENTED** [review]
What does this PR propose for ?

**Comment by Li0k** [comment]
> What does this PR propose for ?

https://github.com/risingwavelabs/risingwave/issues/13879#issuecomment-1898095953

File: src/meta/src/hummock/manager/mod.rs [hunk 1]
```diff
@@ -2692,91 +2692,41 @@ impl HummockManager {
...
+        let (tx, rx) = unbounded_channel();
 
-                    compactor_request_streams.push(future);
-                };
+        let mut join_handle_vec = Vec::default();
 
-            let mut event_loop_iteration_now = Instant::now();
-            loop {
-                let shutdown_rx_shared = shutdown_rx_shared.clone();
+        async fn compact_task_event_handler(
```
Comment (Little-Wallace) [inline src/meta/src/hummock/manager/mod.rs:2706]: How about define this method outside

Comment (Li0k) [reply src/meta/src/hummock/manager/mod.rs]: done

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Review by zwang28 — APPROVED** [review]
LGTM

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

File: src/meta/src/hummock/manager/mod.rs [hunk 2]
```diff
@@ -2744,23 +2753,20 @@ impl HummockManager {
 
                     result = pending_on_none(compactor_request_streams.next()) => {
                         let mut compactor_alive = true;
-                        let (context_id, compactor_stream_req) = result;
+
+                        let (context_id, compactor_stream_req): (_, (std::option::Option<std::result::Result<SubscribeCompactionEventRequest, _>>, _)) = result;
```
[the remark this answers is not in the document]
Comment (Li0k) [reply src/meta/src/hummock/manager/mod.rs:2757]: It required by compiler (

File: src/meta/src/hummock/manager/mod.rs [hunk 3]
```diff
@@ -2744,23 +2753,20 @@ impl HummockManager {
...
-                                hummock_manager.compactor_manager
-                                    .remove_compactor(context_id);
+                                tracing::warn!(error = %err.as_report(), "compactor stream {} poll with err, recv stream may be destroyed", context_id);
                                 continue
                             }
 
                             _ => {
-                                tracing::warn!("compactor {} leaving the cluster", context_id);
-                                hummock_manager.compactor_manager
-                                    .remove_compactor(context_id);
```
[the remark this answers is not in the document]
Comment (Li0k) [reply src/meta/src/hummock/manager/mod.rs:2763]: `remove_compactor ` will remove the `tx` via context_id. Therefore, we only remove tx when send fails and rx when recv fails.

In short: avoid having multiple branches operating removing tx

**Pull request merged**